network/src/dms.rs

```diff
@@ -359,109 +438,170 @@ impl<N: GossipNetwork, S: Storage> DistributedMessageSet<N, S> {
+                    as Arc<dyn DistributedMessageSetRpcInterface>),
+            )]
+            .iter()
+            .cloned()
+            .collect(),
+        )
+        .await;
+        Ok(())
+    }
+
+    async fn serve_fetch(
+        this: Arc<RwLock<Self>>,
         network_config: NetworkConfig,
-        rpc_port: u16,
         peers: SharedKnownPeers,
-    ) -> Result<tokio::task::JoinHandle<Result<(), Error>>, Error> {
-        let mut recv = N::serve(network_config.clone(), peers.clone()).await?;
-        let storage_ = Arc::clone(&self.storage);
-        let peers_ = peers.clone();
-        let network_config_ = network_config.clone();
-        let filter = Arc::clone(&self.filter);
-        let fetch_task = async move {
-            let interval = if let Some(x) = self.config.fetch_interval {
-                x
-            } else {
-                // No fetch task!
-                return Result::<(), Error>::Ok(());
-            };
-            loop {
-                let peers = peers_.read().await;
-                if let Err(e) = fetch(
-                    Arc::clone(&storage_),
-                    &network_config_,
-                    &peers,
-                    filter.as_ref(),
-                )
-                .await
-                {
-                    log::warn!("failed to parse message from the gossip network: {}", e);
-                }
-                tokio::time::sleep(interval).await;
-            }
+    ) -> Result<(), Error> {
+        let interval = if let Some(x) = this.read().await.config.fetch_interval {
+            x
+        } else {
+            return Result::<(), Error>::Ok(());
         };
-        let storage_ = Arc::clone(&self.storage);
-        let peers_ = peers.clone();
-        let broadcast_task = async move {
-            let interval = if let Some(x) = self.config.broadcast_interval {
-                x
-            } else {
-                return Result::<(), Error>::Ok(());
-            };
-            loop {
-                let peers = peers_.read().await;
-                let messages = read_messages(&*storage_.read().await).await?;
-                let tasks = messages.into_iter().map(|message| {
-                    let network_config = network_config.clone();
-                    let peers = peers.clone();
-                    async move {
-                        N::broadcast(
-                            &network_config,
-                            &peers,
-                            serde_json::to_vec(&message).unwrap(),
-                        )
-                        .await?;
-                        Result::<(), Error>::Ok(())
-                    }
-                });
-                join_all(tasks).await;
-                tokio::time::sleep(interval).await;
+        loop {
+            if let Err(e) = Self::fetch(
+                &mut *this.write().await,
+                &network_config,
+                &peers.read().await,
```

Comment:
Don't know how smart Rust is, but does it automatically drops the locks when the fetch is done, in the loop?

Reply:
yes.

Reply:
okok